### ISSUE TITLE
Allow PINODEP008 warning message to be surpressed

### ISSUE THREAD

**Issue by mojavelinux** [pinojs/pino#1201]
The PINODEP008 warning message is blocking me from being able to upgrade to Pino 7. I prefer not to use the new transports feature in my application at this time, and I'm able to stick with pino.destination/SonicBoom for structured log messages. However, when enabling the prettyPrint option, pino emits the PINODEP008 warning message. This message shows up to end users of my application. Otherwise, pino works as expected in this mode. I would like to have the option to suppress this message.

I really don't see the benefit of using transports when using pino-pretty, but these two features seem to have been coupled in pino. If pino-pretty is intended for development, the performance gains transports provide really aren't necessary. The other downside of transports is that we lose the append and mkdir flags that pino.destination/SonicBoom provides.

**Comment by mojavelinux** [comment]
I propose something like:

```js
prettyPrint: {
  suppressDeprecationWarning: true,
}
```

**Comment by mojavelinux** [comment]
I'm willing to submit a PR if this proposal is accepted.

**Comment by jsumners** [comment]
This issue indicates to me that the warning is serving its purpose. We will remove direct integration with `pino-pretty` as that module is now a `pino@7` compatible transport destination. It is unclear to me why you are unable/unwilling to use it as such.

See https://github.com/pinojs/pino/issues/1106

**Comment by mcollina** [comment]
There is no need to undeprecate this. You can continue using pino-pretty with minimal changes, look at https://github.com/pinojs/pino-pretty#programmatic-integration.

**Comment by mojavelinux** [comment]
> jsumners wrote:
> This issue indicates to me that the warning is serving its purpose.

No. I, the developer, got the message. And, having seen it, I'll start preparing to research and make that transition. In the meantime, there's absolutely no reason users of my application should have to see this message as there's nothing they can do about it.

> It is unclear to me why you are unable/unwilling to use it as such.

First of all, it's a deprecation, not a removal. I should be able to continue using this feature as a transition step. I need time to rework my application. That's time I simply don't have right now. The whole purpose of a deprecation is to give me that time. Not to nag my users about it.

> We will remove direct integration with pino-pretty as that module is now a pino@7 compatible transport destination.

I certainly hope you don't remove it until pino@8 as that would be a breaking change. So is it a deprecation or a removal?

As for why I'm unable to use the new transports, I've stated my reasons, but I will be more clear.

To start, this was completely sprung on users of pino. Transports are fundamentally different from destinations in how they are configured and how they work. My application relies on being able to use destinations. So now I have to rework the integration and figure out how to map the configuration in my application onto the pino configuration. It takes time and effort. And you've provided absolutely no opportunity to make that transition. That's not respectful to your users.

The file transport is also missing capabilities that destinations provide:

* mkdir - if true, make the directory for the target file if it doesn't exist
* append - if false, truncate the target file before writing to it
* bufferSize - batch writes to the log file
* sync - write using sync or async operations

I get that perhaps bufferSize and sync are perhaps no longer needed. But the arguments for why they aren't needed aren't stated anywhere I could find.

With destinations, I was able to control the destination file handle so that I could clean up the logger and reopen a new one while the application is running. I don't see how I can do that now.

> mcollina wrote:
> You can continue using pino-pretty with minimal changes, look at pinojs/pino-pretty#programmatic-integration.

I've studied that section, and it's yet another reason why I'm unable to move. If I try to pass the customPrettifiers option, I get a serialization error. So now I have to create a custom pino-pretty transport in a separate file to pass in any options that accept functions. That's not "minimal changes". That's a large change (which I also consider to be a hack).

It's simply not accurate to say that transports are a drop-in replacement for destinations.

> There is no need to undeprecate this.

For reasons stated above, I'm very frustrated with this decision. I think it's very indifferent towards existing users.

**Comment by mcollina** [comment]
Note that you can silence warnings using the [`--no-warnings`](https://nodejs.org/dist/latest-v16.x/docs/api/cli.html#--no-warnings) flag of Node.js. We use the same key mechanism so there is _already_ a crude way to silence the warnings.

You are also not forced to update. Pino@6 stays there and I have not migrated any of my applications yet.

The `pino/file` _is_ `pino.destination()`. It could support the exact same options - we might have forgotten to expose a few and those are likely tiny bugs. Please report them independently or send PRs to fix. 

I don't understand how the deprecation of `pinoPretty: true` relates to destinations at all. All the destination system keep staying there. The goal of this deprecation is to remove the `require('pino-pretty')` from within the internals of pino and a number of hacks we need to support    it.

You can create a pino-pretty stream and pass it as a stream to pino, like so:

```js
const pino = require('pino')
const pretty = require('pino-pretty')
const stream = pretty({
  prettyPrint: { colorize: true }
})
const logger = pino(stream)

logger.info('hi')
```

If you could open issues to flag what is missing specifically it would be great, so we can fill the gaps.

**Comment by mojavelinux** [comment]
> Note that you can silence warnings using the --no-warnings flag of Node.js.

That also requires end users to do something differently, so that's not an option for me.

> The pino/file is pino.destination(). It could support the exact same options - we might have forgotten to expose a few and those are likely tiny bugs. Please report them independently or send PRs to fix.

I will do so.

> I don't understand how the deprecation of pinoPretty: true relates to destinations at all. All the destination system keep staying there.

Because when I use `prettyPrint` (not `pinoPretty` btw) I get a deprecation warning.

> You can create a pino-pretty stream and pass it as a stream to pino

I was not aware I could call it in that way. I will research that to see if it will allow me to continue to use destination + pino-pretty without a warning. I'm not stuck on using the `prettyPrint` option. I just thought that was the only way to get pretty it to wrap a destination. Keep in mind, though, that still requires that I rework my application.

> The goal of this deprecation is to remove the require('pino-pretty') from within the internals of pino and a number of hacks we need to support it.

That's a very reasonable goal. Ultimately, what I need to match the existing behavior of my application is pino-pretty + destination, not necessarily the `prettyPrint` option.

> You are also not forced to update. Pino@6 stays there and I have not migrated any of my applications yet.

The whole point is that I'm trying to upgrade so my application doesn't fall behind. And it's the only library I cannot upgrade right now.

**Comment by mojavelinux** [comment]
> You can create a pino-pretty stream and pass it as a stream to pino

There seems to be something preventing this from working. When I use the following destination:

```js
pretty({
  destination: new SonicBoom({ dest: '/path/to/log.txt', mkdir: true, append: false }),
})
```

I get:

```
Error: SonicBoom supports only file descriptors and files
```

The internal transport seems to be tainting the SonicBoom instance somehow.

**Comment by mojavelinux** [comment]
There seems to be quite a lot of patching in pino to get pino-pretty to function properly as a destination. Here's the only way I could figure out how to do it:

```js
const pino = require('pino')
const { getPrettyStream } = require('pino/lib/tools')

const logger = pino({
  level: 'info',
}, getPrettyStream({
  colorize: false,
  translateTime: 'SYS:HH:MM:ss.l',
  customPrettifiers: {
    foo: () => 'replaced!',
  },
}, undefined, new SonicBoom({ dest: '/tmp/path/to/log.txt', mkdir: true, append: false })))
logger.replace({ foo: 'bar' }, 'hi')
```

I'd be curious to know how to use the pino-pretty function reliably in place of getPrettyStream.

**Comment by jsumners** [comment]
> First of all, it's a deprecation, not a removal. I should be able to continue using this feature as a transition step. I need time to rework my application. That's time I simply don't have right now. The whole purpose of a deprecation is to give me that time. Not to nag my users about it.

I said we _will_ remove it in the next major, not that it is removed, or being removed, in the current major release. That is the _purpose_ of the warning to give you time to deal with it. If you don't like the warnings, use `--no-warnings` or set `NODE_NO_WARNINGS=1` in your environment.

> To start, this was completely sprung on users of pino. Transports are fundamentally different from destinations in how they are configured and how they work. My application relies on being able to use destinations. So now I have to rework the integration and figure out how to map the configuration in my application onto the pino configuration. It takes time and effort. And you've provided absolutely no opportunity to make that transition. That's not respectful to your users.

We didn't "spring" anything on anyone. We worked in the open. We issued several release candidates, advertising them on our social channels, asking for feedback. And we issued a _semver major_. We _did not_ completely change the API. I would be wholly on your side if that were the case. The API evolution is quite incremental.

> For reasons stated above, I'm very frustrated with this decision. I think it's very indifferent towards existing users.

Again, we worked in the open and released a major version according to the contract with our users.

**Comment by mojavelinux** [comment]
Given that you're unwilling to add a simple flag to control the deprecation warning, I respectfully disagree that you're providing a reasonable contract with users. This is not a smooth transition, and I have clearly cited why that is the case.

**Comment by mojavelinux** [comment]
> We didn't "spring" anything on anyone.

Yes, you did. You added a new mechanism for log targets (transports) and deprecated the existing prettyPrint integration with a warning that can't be removed programmatically, all in the same major version. That's springing a change on users. The pino@7 release would have been the right opportunity to communicate a major change so that users could begin transitioning for pino@8.

It's not acceptable, in my mind, to have to ask my users make a change so they don't get a warning they cannot control or for me to globally disable warnings in Node. It's an unreasonable request. And I can't transition yet to transports because they simply aren't ready to meet existing functionality. So I'm stuck on pino@6 if I want to avoid a warning.

**Comment by jsumners** [comment]
> You can disable it by adding one line as the first line in the entry point to your application:

I stand corrected. Just tried it with the following and the warning shows 😕

```js
process.env.NODE_NO_WARNINGS = 1
const log = require('pino')({ prettyPrint: true })
log.info('blah')
```

**Comment by mcollina** [comment]
It's now possible to pass custom sonic-boom instances as destination to pino-pretty stream factory. This will not work through `pino.transport()`.

**Comment by mojavelinux** [comment]
I'm able to verify I can configure pino pretty as a destination.

```js
const logger = pino({
  level: 'info',
}, pretty({
  destination: new SonicBoom({ dest: '/tmp/path/to/log.txt', mkdir: true, append: false }),
  colorize: true,
  translateTime: 'SYS:HH:MM:ss.l',
  customPrettifiers: {
    foo: () => 'replaced!',
  },
}))
```

Thanks for applying that fix.

I'm now convinced that this arrangement is better than the deprecation warning as it keeps the two concerns separate. Thanks again.

I will still follow-up with the support for mkdir and append on transports, but that's now a parallel concern.

**Comment by mojavelinux** [comment]
I have successfully integrated pino@7 into my application with very little change...and the right change at that. Thank you again Matteo and James for helping to ease the transition after my initial report. I'm grateful for the work that you do.

I have filed the follow-up issues about the append and mkdir options, as promised. See #1221 and #1222. I will submit matching PRs for your review. Fortunately, the implementation seems to be very straightforward.
